ReplayTables/ReplayBuffer.py

```diff
@@ -1,118 +1,68 @@
 import numpy as np
 from abc import abstractmethod
-from typing import Any, Dict, Generic, Iterable, List, NewType, Tuple, TypeVar, Type, Union, cast
+from typing import Any, Tuple
 from ReplayTables.Distributions import UniformDistribution
-
-T = TypeVar('T', bound=Iterable)
-EID = NewType('EID', int)
-EIDS = NewType('EIDS', np.ndarray)
-
-class ReplayBufferInterface(Generic[T]):
-    def __init__(self, max_size: int, structure: Type[T], rng: np.random.Generator):
+from ReplayTables.interface import Timestep, Batch, EID, EIDs
+from ReplayTables.ingress.IndexMapper import IndexMapper
+from ReplayTables.ingress.CircularMapper import CircularMapper
+from ReplayTables.storage.BasicStorage import BasicStorage
+from ReplayTables.storage.Storage import Storage
+from ReplayTables._utils.LagBuffer import LagBuffer
+
+class ReplayBufferInterface:
+    def __init__(self, max_size: int, lag: int, rng: np.random.Generator):
         self._max_size = max_size
-        self._structure = cast(Any, structure)
+        self._lag = lag
         self._rng = rng
 
-        self._t = 0
-        self._storage: Dict[EID, T] = {}
-
-        self._views: List[ReplayViewInterface[T]] = []
+        self._lag_buffer = LagBuffer[Tuple[EID, Timestep, Any]](maxlen=lag)
+        self._idx_mapper: IndexMapper = CircularMapper(max_size + lag)
+        self._storage: Storage = BasicStorage(max_size + lag, self._idx_mapper)
```

Comment:
This coupling between the storage and idx_mapper is bad. If I want to change out either mechanism, it would require rebuilding _both_ objects and invalidating both.

It would be better if the `eid` -> `idx` mapping is always done externally, then all other objects only have an `idx` api.